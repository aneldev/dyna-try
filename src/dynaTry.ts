import {
  dynaError,
  IDynaError,
} from "dyna-error";

export const dynaTry = <TResolve>(
  {
    try: try_,
    timeout,
    timeoutError = dynaError({
      code: 600408,
      message: `Try timed out (${timeout}ms)`,
    }),
  }: {
    try: () => Promise<TResolve>;
    timeout: number;
    timeoutError?: IDynaError | Error;
  },
): Promise<TResolve> => {
  return new Promise((resolve, reject) => {
    let isResolved = false;
    let isRejected = false;

    const timer = setTimeout(() => {
      if (isResolved || isRejected) return;
      isRejected = true;
      try {
        reject(timeoutError);
      }
      catch (e) {
        reject(e);
      }
    }, timeout);

    try_()
      .then((d) => {
        clearTimeout(timer);
        if (isRejected) return;
        isResolved = true;
        resolve(d);
      })
      .catch((e) => {
        clearTimeout(timer);
        if (isRejected) return;
        isRejected = true;
        reject(e);
      });

  });
};
