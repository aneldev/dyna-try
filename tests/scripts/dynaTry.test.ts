import {dynaTry} from "../../src";
import {dynaError} from "dyna-error";

interface IPerson {
  name: string;
  age: number;
}

describe('dynaTry', () => {

  const getPerson = (id: string, delay: number, _test_willFail: boolean = false): Promise<IPerson> => {
    id;
    return new Promise((resolve, reject) => {
      setTimeout(() => {
        if (_test_willFail) {
          reject(new Error("Server error"));
          return;
        }
        resolve({
          name: 'John Smith',
          age: 34,
        });
      }, delay);
    });
  };

  test('Success try', async () => {
    const person = await dynaTry<IPerson>({
      timeout: 1000,
      try: () => getPerson('232', 500),
    });
    expect(person).toMatchSnapshot();
  });

  test('With error', async () => {
    expect.assertions(2);
    try {
      const person = await dynaTry<IPerson>({
        timeout: 1000,
        try: () => getPerson('232', 500, true),
        timeoutError: dynaError({
          code: 600408,
          message: 'Time out',
        }),
      });
      expect(person).toMatchSnapshot();
    }
    catch (e) {
      expect(e.code).not.toBe(600408);
      expect(e.message).toBe('Server error');
    }
  });

  test('With JS error', async () => {
    expect.assertions(2);
    try {
      const person = await dynaTry<IPerson>({
        timeout: 1000,
        try: () => getPerson('232', 500, true),
        timeoutError: new Error('Time out'),
      });
      expect(person).toMatchSnapshot();
    }
    catch (e) {
      expect(e.code).not.toBe(600408);
      expect(e.message).toBe('Server error');
    }
  });

  test('Timeout', async () => {
    expect.assertions(1);
    try {
      const person = await dynaTry<IPerson>({
        timeout: 1000,
        try: () => getPerson('232', 500),
        timeoutError: dynaError({
          code: 600408,
          message: 'Time out',
        }),
      });
      expect(person).toMatchSnapshot();
    }
    catch (e) {
      expect(e.code).toBe(600408);
    }
  });
});
