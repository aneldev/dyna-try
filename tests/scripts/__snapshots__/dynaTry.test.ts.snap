// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dynaTry Success try 1`] = `
Object {
  "age": 34,
  "name": "John Smith",
}
`;

exports[`dynaTry Timeout 1`] = `
Object {
  "age": 34,
  "name": "John Smith",
}
`;

exports[`dynaTry With error 1`] = `
Object {
  "age": 34,
  "name": "John Smith",
}
`;
